Corona Product

alifa98 opened the pull request:
Hi all,

I have added the "corona product" of two graphs in the product algorithms. 
I also had written a [blog post here](https://blog.alifaraji.ir/math/graph-theory/corona-product.html) about this product months ago.

Thank you

dschult commented:
Thank you for this PR.   The Corona product is a helpful feature to add. And this implementation looks good to me.

Can you add `corona_product` to the list of functuons to link to in the documentation? `doc/reference/algorithms/operators.rst`  That will link this doc_string to the reference section of the docs.

Are there performance issues we should think about?  I think CPU usage is not an issue. What about memory?  If you have a really big H (or G) will there be a bottleneck in new_edges getting huge? If so, you could loop over `list(GH)` and add the nodes/edges to GH as you go.  Anything else like that?

Thanks!

alifa98 commented:
Hi, 
As you said, I have added `corona_product` to `doc/reference/algorithms/operators.rst`.

Your concern about memory usage in large graphs was correct, so I changed my implementation a little bit to avoid temp lists.

Thank you.

dschult commented on networkx/algorithms/operators/product.py:
```diff
@@ -508,22 +508,17 @@ def corona_product(G, H):
 
         # copy nodes of H in GH, call it H_i
-        new_nodes.extend([(G_node, v) for v in H])
+        GH.add_nodes_from([(G_node, v) for v in H])
```
You can even go one step further by removing the square brackets in the function call. That creates a "generator expression" that avoids creating the list. It just yields one node (or edge) at a time as needed.

`GH.add_nodes_from((G_node, v) for v in H)`

dschult commented on networkx/algorithms/operators/product.py:
```diff
@@ -459,3 +460,65 @@ def rooted_product(G, H, root):
+def corona_product(G, H):
+    r"""Returns the Corona product f G and H.
+
+
+        The corona product of $G$ and $H$ is the graph $C = G \circ H$ obtained by
+        taking one copy of $G$, called the center graph, $|V(G)|$ copies of $H$, called the outer graph,
+        and making the $i$-th vertex of $G$ adjacent to every vertex of the $i$-th copy of $H$,
+        where $1 ≤ i ≤ |V (G)|$.
+
+        Parameters
+        ----------
+        G, H: graphs
+         Networkx graphs.
+
+        Returns
+        -------
+        C: NetworkX graph
+         The Corona product of G and H.
+
+        Raises
+        ------
+        NetworkXError
+         If G and H are not both directed or both undirected.
+
+        References
+        ----------
+        [1] M. Tavakoli, F. Rahbarnia, and A. R. Ashrafi, “Studying the corona product of graphs under some graph invariants,” Transactions on Combinatorics, vol. 3, no. 3, pp. 43–49, Sep. 2014, doi: 10.22108/toc.2014.5542.
+        [2] A. Faraji, “Corona Product in Graph Theory,” Ali Faraji, May 11, 2021. https://blog.alifaraji.ir/math/graph-theory/corona-product.html (accessed Dec. 07, 2021).
+    ‌
+        Examples
+        --------
+        >>> G = nx.cycle_graph(6)
+        >>> H = nx.path_graph(2)
+        >>> C = nx.algorithms.product.corona_product(G, H)
+        >>> list(C)
+        [0, 1, 2, 3, 4, 5, (0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1), (3, 0), (3, 1), (4, 0), (4, 1), (5, 0), (5, 1)]
+        >>> nx.info(C)
+        'Graph with 18 nodes and 24 edges'
```
Some picky changes to shorten the lines in the doc_strings:
```suggestion

        The corona product of $G$ and $H$ is the graph $C = G \circ H$ obtained by
        taking one copy of $G$, called the center graph, $|V(G)|$ copies of $H$,
        called the outer graph, and making the $i$-th vertex of $G$ adjacent to 
        every vertex of the $i$-th copy of $H$, where $1 ≤ i ≤ |V (G)|$.

        Parameters
        ----------
        G, H: NetworkX graphs
            The graphs to take the carona product of.

        Returns
        -------
        C: NetworkX graph
            The Corona product of G and H.

        Raises
        ------
        NetworkXError
            If G and H are not both directed or both undirected.

        References
        ----------
        [1] M. Tavakoli, F. Rahbarnia, and A. R. Ashrafi, 
            “Studying the corona product of graphs under some graph invariants,” 
            Transactions on Combinatorics, vol. 3, no. 3, pp. 43–49, Sep. 2014, 
            doi: 10.22108/toc.2014.5542.
        [2] A. Faraji, “Corona Product in Graph Theory,” Ali Faraji, May 11, 2021. 
            https://blog.alifaraji.ir/math/graph-theory/corona-product.html (accessed Dec. 07, 2021).
    ‌
        Examples
        --------
        >>> G = nx.cycle_graph(4)
        >>> H = nx.path_graph(2)
        >>> C = nx.corona_product(G, H)
        >>> list(C)
        [0, 1, 2, 3, (0, 0), (0, 1), (1, 0), (1, 1), (2, 0), (2, 1), (3, 0), (3, 1)]
        >>> print(C)
        'Graph with 12 nodes and 16 edges'
```

alifa98 replied:
thank you!

alifa98 commented:
> oops... my suggestion had quotes around the output string when it shouldn't have. I think taking out those quotes gets rid of the test failures.

I do not know what to do!!!?!

dschult commented:
Take the single quotes out of line 508 where the output of the example appears as `'Graph with 12 nodes and 16 edges'` when it should be `Graph with 12 nodes and 16 edges`

dschult commented on networkx/algorithms/operators/product.py:
```diff
@@ -459,3 +460,69 @@ def rooted_product(G, H, root):
+@not_implemented_for("directed")
+@not_implemented_for("multigraph")
+def corona_product(G, H):
+    r"""Returns the Corona product f G and H.
+
+
+        The corona product of $G$ and $H$ is the graph $C = G \circ H$ obtained by
+        taking one copy of $G$, called the center graph, $|V(G)|$ copies of $H$,
+        called the outer graph, and making the $i$-th vertex of $G$ adjacent to
+        every vertex of the $i$-th copy of $H$, where $1 ≤ i ≤ |V (G)|$.
```
```suggestion
    r"""Returns the Corona product of G and H.

        The corona product of $G$ and $H$ is the graph $C = G \circ H$ obtained by
        taking one copy of $G$, called the center graph, $|V(G)|$ copies of $H$,
        called the outer graph, and making the $i$-th vertex of $G$ adjacent to
        every vertex of the $i$-th copy of $H$, where $1 ≤ i ≤ |V(G)|$.
```

dschult commented on networkx/algorithms/operators/product.py:
```diff
@@ -459,3 +460,69 @@ def rooted_product(G, H, root):
+        Parameters
+        ----------
+        G, H: NetworkX graphs
+            The graphs to take the carona product of.
+
```
```suggestion
            The graphs to take the carona product of.
            `G` is the center graph and `H` is the outer graph

```

dschult commented on networkx/algorithms/operators/tests/test_product.py:
```diff
@@ -424,3 +424,13 @@ def test_rooted_product():
+    H = nx.path_graph(2)
+    C = nx.corona_product(G, H)
+    assert len(C) == (len(G) * len(H)) + len(G)
+    assert C.number_of_edges() == G.number_of_edges() + len(
+        G
+    ) * H.number_of_edges() + len(G) * len(H)
```
```suggestion
    assert C.size() == G.size() + len(G) * H.size() + len(G) * len(H)
```

dschult reviewed (commented):
I have some minor typos in the docs, an extra line to name G and H, and a shorter assert line in the test (which matches the previous test's use of `G.size()` over `G.number_of_edges()`.

With these changes I'm ready to approve it.  Thanks very much!!

dschult reviewed (approved):
Thanks!

MridulS reviewed (approved):
Thanks @alifa98!
